Ray IPv6 support

samrocketman opened the pull request:
Why are these changes needed?
-----------------------------

Provides support for IPv6 only and dual-stack IPv4/IPv6.

Inspired by this discussion https://discuss.ray.io/t/lack-of-ipv6-support/2323

IPv6 is much more common than it used to be; in some cases the only available stack is IPv6.  Commonly machines come with dual-stack connectivity enabled (IPv4 and IPv6 available at the same time).

I've already reviewed the source code of [kuberay][kuberay].  I've determined due to its use of native Go net methods kuberay already supports IPv6.

Design
------

- See `python/ray/_private/net.py`. Centralize networking code to make it easier to support IPv6 alongside IPv4.  This is simpler and clarifies where in your network code you need to support this when using low level python sockets.
- Favor IPv4 since that is what this project is typically used to; code will fall back to IPv6 if IPv4 is not available.
- It uses dual-stack IPv4 and IPv6 when it is available; for example when listening on a port across all interfaces or determining free ports.
- All places I could find for string parsing IP or host and port have been substituted.  With a central method, updates can be made from which all dependent areas will benefit.

Shell code
----------

I notice the following scripts may or may not work with IPv6.  They should probably updated using [bash parameter expansion][bash-pe].

* `doc/source/cluster/doc_code/slurm-basic.sh`
* `doc/source/cluster/doc_code/slurm-template.sh`

If I update them; it will be in a separate PR.

For example, the following bash code properly parses both IPv4 and IPv6 `ip:port` into two parts.

    # IPv6 loopback port 8080
    ip_and_port=::1:8080
    ip="${ip_and_port%:*}"
    port="${ip_and_port##*:}"

    # IPv4 loopback port 8080
    ip_and_port=127.0.0.1:8080
    ip="${ip_and_port%:*}"
    port="${ip_and_port##*:}"

Related issue number
--------------------

I don't know if it will close all of these issues but I did an issues search for `ipv6`.

* closes #6967
* #43416
* #41966
* #43910

Duplicated by

* https://github.com/ray-project/ray/pull/40332

Checks
------

Initially, this pull request is in-draft.  This is to enable reviewing sooner and manual testing to occur.  Automated tests to follow.

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I added a method in Tune, I've added it in `doc/source/tune/api/` under the corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

[bash-pe]: https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
[kuberay]: https://github.com/ray-project/kuberay

Commit:
Reuse C++ changes from PR #40332

This is the literal changes from @jjacobj84 from while I will build upon in later commits.

Specifically files:

- src/ray/gcs/gcs_client/gcs_client.h
- src/ray/gcs/pubsub/gcs_pub_sub.cc
- src/ray/rpc/grpc_client.h

Sam Gleske (samrocketman) has updated these files resolving conflicts
but left them in a similar form Japson originally contributed.

See also
--------

https://github.com/ray-project/ray/pull/40332

Signed-off-by: Japson Jeyasekaran <jjacobj@amazon.com>
Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sam.mxracer@gmail.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman committed:
[Core] Ray IPv6 support

Why are these changes needed?
-----------------------------

Provides support for IPv6 only and dual-stack IPv4/IPv6.

Inspired by this discussion https://discuss.ray.io/t/lack-of-ipv6-support/2323

IPv6 is much more common than it used to be; in some cases the only available stack is IPv6.  Commonly machines come with dual-stack connectivity enabled (IPv4 and IPv6 available at the same time).

I've already reviewed the source code of [kuberay][kuberay].  I've determined due to its use of native Go net methods kuberay already supports IPv6.

Design
------

- Centralize networking code to make it easier to support IPv6 alongside IPv4.  This is simpler and clarifies where in your network code you need to support this when using low level python sockets.
- Favor IPv4 since that is what this project is typically used to; code will fall back to IPv6 if IPv4 is not available.
- It uses dual-stack IPv4 and IPv6 when it is available; for example when listening on a port across all interfaces.

Shell code
----------

I notice the following scripts may or may not work with IPv6.  They should probably updated using [bash parameter expansion][bash-pe].

* `doc/source/cluster/doc_code/slurm-basic.sh`
* `doc/source/cluster/doc_code/slurm-template.sh`

If I update them; it will be in a separate PR.

For example, the following bash code properly parses both IPv4 and IPv6 `ip:port` into two parts.

    # IPv6 loopback port 8080
    ip_and_port=::1:8080
    ip="${ip_and_port%:*}"
    port="${ip_and_port##*:}"

    # IPv4 loopback port 8080
    ip_and_port=127.0.0.1:8080
    ip="${ip_and_port%:*}"
    port="${ip_and_port##*:}"

Related issue number
--------------------

I don't know if it will close all of these issues but I did an issues search for `ipv6`.

* closes #6967
* #43416
* #41966
* #43910

Checks
------

Initially, this pull request is in-draft.  This is to enable reviewing sooner and manual testing to occur.  Automated tests to follow.

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

[bash-pe]: https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
[kuberay]: https://github.com/ray-project/kuberay

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>

samrocketman committed:
Address self-code review

* Fixed a few function calls missing net import.
* Missed ip parse
* Fix net.py syntax
* Fix tls_utils.py syntax

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>

samrocketman commented on python/ray/util/rpdb.py:
```diff
@@ -102,7 +103,7 @@ def __init__(
         self._quiet = quiet
         self._patch_stdstreams = patch_stdstreams
-        self._listen_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._listen_socket = net._get_sock_stream_from_host(host)
```
Make this dual-stack capable.

samrocketman commented on python/ray/util/rpdb.py:
```diff
@@ -336,10 +337,9 @@ def _post_mortem():
     )
     rdb.post_mortem()
 
-
```
Fix this removed line; by mistake

samrocketman commented on python/ray/util/rpdb.py:
```diff
@@ -336,10 +337,9 @@ def _post_mortem():
-    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    s.connect((host, port))
+    s = net._get_sock_stream_from_host(host)
+    s.connect((address, port))
```
Should be `host`, not `address`.

samrocketman commented on python/ray/workflow/tests/conftest.py:
```diff
@@ -39,7 +39,7 @@ def _workflow_start(storage_url, shared, use_ray_client, **kwargs):
     with _init_cluster(storage_url, **kwargs) as cluster:
         if use_ray_client == "ray_client":
-            address = f"ray://{cluster.address.split(':')[0]}:10001"
+            address = f"ray://{cluster.address.rsplit(":", 1)[0]}:10001"
```
Use net ip parse.

samrocketman committed:
Fix net.py

Misuse of dualstack.  It should only be used when listening for a port on
all interfaces.

flake8 fixes

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>

samrocketman commented on python/ray/_private/tls_utils.py:
```diff
@@ -33,19 +33,26 @@ def generate_self_signed_tls_certs():
+        private_dns_addresses.append(
+            x509.DNSName(socket.gethostbyname(socket.gethostname()))
+        )
+    except:
```
```suggestion
    except Exception as ignored:  # noqa: F841
```

samrocketman commented on python/ray/serve/_private/proxy.py:
```diff
@@ -1327,7 +1328,7 @@ async def ready(self):
 
     async def run_http_server(self):
-        sock = socket.socket()
+        sock = net._get_socket_dualstack_fallback_single_stack_laddr()
```
```suggestion
        sock = net._get_sock_stream_from_host(self.host)
```

samrocketman commented on python/ray/tests/conftest.py:
```diff
@@ -997,15 +998,16 @@ def set_runtime_env_retry_times(request):
         for i in range(MAX_RETRY):
             try:
-                sock.bind(("127.0.0.1", port))
+                # Binds to all loopback interfaces (IPv4, or IPv6, or both if dualstack support)
```
```suggestion
```

samrocketman commented on python/ray/tests/conftest.py:
```diff
@@ -997,15 +998,16 @@ def set_runtime_env_retry_times(request):
     port = getattr(request, "param", 0)
     try:
-        sock = socket.socket()
+        sock = net._get_socket_dualstack_fallback_single_stack_laddr()
```
```suggestion
        sock = net._get_sock_stream_from_host("localhost")
```

samrocketman commented on python/ray/util/rpdb.py:
```diff
@@ -102,7 +103,7 @@ def __init__(
         self._quiet = quiet
         self._patch_stdstreams = patch_stdstreams
-        self._listen_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._listen_socket = net._get_socket_dualstack_fallback_single_stack_laddr()
```
```suggestion
        self._listen_socket = net._get_sock_stream_from_host(host)
```

samrocketman reviewed (commented):
Another review pass fixing issues I learned studying how Python sockets support IPv6.

I thought dualstack works when binding with a host but it does not (e.g. localhost dualstack only binds to IPv6 `::1` but not `127.0.0.1`).

Binding dualstack only works when binding a port to all interfaces i.e. `sock.bind(('', port))`.

samrocketman committed:
Fix additional network issues

- Bugfix python/ray/util/client/server/proxier.py not obtaining local_ip address from localhost.
- Remove use of staticmethod; the way I used it isn't compatible below Python 3.10
- Supprt parsing IPv6 URL format `proto://[ip]:port/` in net.py
- Fixed API linting issues in net.py
- Fix pylint issues

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+from contextlib import closing
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+    ]
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+        return _get_addrinfo_from_sock_kind(address, kind)
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+    return socket.socket(inet_address[0], kind)
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+    return _get_sock_from_host(address, socket.SOCK_STREAM)
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+    return _get_sock_from_host(address, socket.SOCK_DGRAM)
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+    return private_ip_addresses
+
+
+@staticmethod
```
```suggestion
```

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,164 @@
+        return _get_sock_from_host("localhost", kind)
+
+
+@staticmethod
```
```suggestion
```

samrocketman reviewed (commented):
I appear to be misusing `@staticmethod`.

samrocketman commented:
I guess my issues with staticmethod are best described here https://stackoverflow.com/questions/1859959/static-methods-how-to-call-a-method-from-another-method

samrocketman commented:
After messing around with `@staticmethod` across different versions of python; the way I used it (on a file of functions without a class definition) worked in Python 3.10 but not Python 3.9.  Rather than trying to declare the methods static which would work across multiple versions of python I decided to remove the symbol from the private net module.

samrocketman commented on python/ray/util/client/server/proxier.py:
```diff
@@ -199,11 +200,12 @@ def create_specific_server(self, client_id: str) -> SpecificServer:
                 self.servers.get(client_id) is None
             ), f"Server already created for Client: {client_id}"
             port = self._get_unused_port()
+            local_ip = net._get_addrinfo_from_sock_kind_ipv4_fallback_ipv6("localhost", socket.SOCK_STREAM)
```
```suggestion
            local_ip = net._get_addrinfo_from_sock_kind_ipv4_fallback_ipv6("localhost", socket.SOCK_STREAM)[1]
```

samrocketman commented on dashboard/tests/test_dashboard.py:
```diff
@@ -1059,7 +1059,7 @@ def test_agent_port_conflict(shutdown_only):
 
     # ocuppy the port with a socket.
-    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s = net._get_socket_dualstack_fallback_single_stack_laddr()
```
```suggestion
    s = net._get_sock_stream_from_host("localhost")
```

samrocketman committed:
Update dashboard/tests/test_dashboard.py

Signed-off-by: Sam Gleske <875669+samrocketman@users.noreply.github.com>

samrocketman committed:
Minor fixes for tests

- More pylint changes.
- Exception handling added for existing tests.

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>

samrocketman committed:
Bugfix type coercion

Some methods pass bytes instead of strings when parsing IP and port.  I
added extra type safety instead of assuming.

Fixed some pylint issues.

Trying to use net class in pxi.

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>

samrocketman commented:
This is ready for general testing.  All of the automated tests pass except for linting and I somehow need to account for `unittest.mock.MagicMock` type being passed to the IP address parser.

This is pretty close to leaving draft.  I need to write some automated tests.

samrocketman commented:
Pinging @rkooo567 @jjacobj84 for visibility.  I appear to have duplicated some effort and I plan to keep iterating on this PR until it is ready.

samrocketman commented:
From the failure 

```
[2024-03-25T03:13:33Z] E           ValueError: address type not str or bytes:
[2024-03-25T03:13:33Z] E
[2024-03-25T03:13:33Z] E               type: <class 'unittest.mock.MagicMock'>
[2024-03-25T03:13:33Z] E              value: <MagicMock name='mock.gcs_address' id='[140017679483136](tel:140017679483136)'>
```

I don’t want to handle mock classes in application code.  How should I handle this?

anyscalesam commented:
@samrocketman - we'll look for opp to pair up someone here to help this through; stay tuned next Monday~

samrocketman commented:
@anyscalesam sounds good.  I am available to continue working on this just holding per my questions; I know there’s code style issues but I’ll fix them before taking out of draft.

samrocketman commented:
I posted in https://github.com/ray-project/ray/pull/40332 asking author if they mind if I pull in their C/C++ changes.  If they don’t answer then I’ll update these files myself with my own take.

anyscalesam commented:
been a week @samrocketman > I think safe to pick; just include the OG author and/or reference it in your PR? Let's not wait. My two cents (feel free to disagree)

anyscalesam commented:
@samrocketman not pushing but any ETA on when you think you'll open this up for review?

samrocketman commented on python/ray/_private/net.py:
```diff
@@ -0,0 +1,177 @@
+    if isinstance(address, bytes):
+        address = address.decode('ascii')
+
+    if not isinstance(address, str):
+        raise ValueError(
+            "address type not str or bytes:\n\n" +
+            f"    type: {repr(type(address))}\n" +
+            f"   value: {repr(address)}"
+        )
```
This should probably commented out because of the mock interfering during unit tests.

samrocketman commented:
@anyscalesam worth calling out I've not operated Ray before.  I've been making code changes based on core networking and programming concepts/experience.

So when this is ready/out of draft any help in testing this in an ipv6 environment would be appreciated.

dsarfati commented:
The ability to run in an ipv6 only network is also a blocker for us. I would be happy to try and run some tests as well so we can get this merged.

semoal commented:
Friendly ping here :)

csingley commented:
So I rebased this PR onto master, built the Docker image, and deployed it onto an IPv6-only Kuberentes cluster via KubeRay.  This cleared several errors in the Python code and allowed GCS to start... but alas, there still appears to be more IPv4-related tech debt lurking within the CPP source, as revealed by [GCS logs](https://gist.github.com/csingley/1c9e6209537ae3dc87a22a58d8e8d4aa) when trying to connect to Redis:

```
2024-08-28 08:09:47,248 I 75 75] (gcs_server) gcs_server.cc:518: Using external Redis for KV storage: redis:6379
[2024-08-28 08:09:47,278 I 75 75] (gcs_server) gcs_server.cc:73: GCS storage type is StorageType::REDIS_PERSIST
[2024-08-28 08:09:47,279 I 75 75] (gcs_server) redis_context.cc:474: Resolve Redis address to fdf5:6da1:fe0d:cc1e::5a9c
[2024-08-28 08:09:47,279 I 75 75] (gcs_server) redis_context.cc:350: Attempting to connect to address fdf5:6da1:fe0d:cc1e::5a9c:6379.
```

As best as I can determine, the root cause appears to be traceable to [`RedisAsioClient`](https://github.com/ray-project/ray/blob/master/src/ray/gcs/asio.cc#L51).

Searching the source code for occurrences of the string `"v4"` also turns up `network_util.cc` - specifically, problematic code in [CheckFree](https://github.com/ray-project/ray/blob/master/src/ray/common/network_util.cc#L22-L30) , which seems likely to break the Raylet [`WorkerPool`](https://github.com/ray-project/ray/blob/master/src/ray/raylet/worker_pool.cc#L661).  This problem also appears in [a](https://github.com/ray-project/ray/blob/master/src/ray/raylet/runtime_env_agent_client_test.cc#L47-L48) [few](https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_client/test/gcs_client_reconnection_test.cc#L131) [tests](https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/test/gcs_health_check_manager_test.cc#L42-L43).

Searching the source code for occurrences of the strings `"127.0.0.1"` and `"0.0.0.0"` also yields several hits - mostly in various tests, but also in (e.g.)  [`GrpcServer`](https://github.com/ray-project/ray/blob/master/src/ray/rpc/grpc_server.cc#L61-L62), [`ObjectManager`](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_manager.cc#L100-L104), and [`NodeManager`](https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L241-L243).

These bugs are blockers to operating Ray within IPv6-only environments.

anyscalesam commented:
thanks for doing a pass here @csingley; we'll take a look and get back to you on next steps here

aspexdaniel commented:
this also becomes a blocker for us after we switched to an ipv6-only k8s environment.

anyscalesam commented:
TY for the cont'd feedback; keep it coming. It helps us with the right data signal to prioritize right projects.

samrocketman commented:
I'll look at picking this up again. I made these changes in concept only (I don't currently operate ray but my environment does have ipv6 only clusters).

aspexdaniel commented:
I can help to test out if we have a testing image for arm64. Thanks for your effort.

xtuc commented:
@anyscalesam We at Cloudflare would also benefit from IPv6 support.

anyscalesam commented:
@xtuc @aspexdaniel @samrocketman can y'all reach out to me on ray.slack.com so I can invite you to the Ray Contributors sync to get some more traction on this?

I can connect you with the Ray Core on-call so we can get a plan of attack together to get this pushed through. Pretty good improvement for the community so I feel well worth doing.

samrocketman commented:
> ... can y'all reach out to me on ray.slack.com so I can invite you to the Ray Contributors sync to get some more traction on this?


It says I do not have an account and I don't see signup links on the ray website.

samrocketman commented:
I think it makes most sense to do this in two parts:

1. These supporting python changes which DRY up socket usage.  The intent here is the code was reorganized a little bit and we want to make sure that existing ray functionality works well.
2. c++ changes is the last bit of change for IPv6 support.  I'm comfortable making the cpp changes but it might make most sense to do it as a separate PR once it is verified the python changes are acceptable.

aspexdaniel commented:
> @xtuc @aspexdaniel @samrocketman can y'all reach out to me on ray.slack.com so I can invite you to the Ray Contributors sync to get some more traction on this?
> 
> I can connect you with the Ray Core on-call so we can get a plan of attack together to get this pushed through. Pretty good improvement for the community so I feel well worth doing.

me too can't find a link to join the slack group.

aspexdaniel commented:
> I think it makes most sense to do this in two parts:
> 
> 1. These supporting python changes which DRY up socket usage.  The intent here is the code was reorganized a little bit and we want to make sure that existing ray functionality works well.
> 2. c++ changes is the last bit of change for IPv6 support.  I'm comfortable making the cpp changes but it might make most sense to do it as a separate PR once it is verified the python changes are acceptable.

as the python codes relies on the c++ changes, will it be possible to submit and test the cpp changes first?

samrocketman commented:
> as the python codes relies on the c++ changes, will it be possible to submit and test the cpp changes first?

My intent was to basically DRY up the IPv4/IPv6 network code (without full support for IPv6).  The goal being get it nicely organized and ensure I didn't break existing functionality.  Then a followup change with the cpp changes with any remaining python changes.

I'm going to take a deeper look at this today I received permission from my place of work to take a look at this during working hours.

samrocketman commented:
Because I have approval to work on this during work hours I'll aim for full IPv6 support in this PR rather than breaking it up over multiple PRs.  My primary blocker up to this point is lack of time in off-hours to work on it (I have several projects I contribute to so it's always a bit of a dice roll which ones I update).

samrocketman commented:
@anyscalesam if you want to invite me to your ray slack instance you can find my email in my commit authorship of this PR

jadams41 commented:
We're interested in this support @ Netflix as well (we're also going the way of IPv6-only for k8s). Would be happy to engage on slack if we could be of any help

samrocketman committed:
Resolve conflicts: Merge remote-tracking branch 'upstream/master' into ipv6

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman commented:
I have resolved the few dozen conflicts with upstream.

Next steps:

* I'm going to do another pass over the python code to make sure the latest HEAD is IPv6-compatible.
* I'm going to follow that up with C or C++ changes which would be required for IPv6.

samrocketman committed:
Update added socket code

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman committed:
Update Java code for IPv6 parsing

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman committed:
Update missing Python changes for IP:port parsing

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman committed:
Fix linter issues

Run Python 3.9 docker container in a safe way.

docker run -u 1000:1000 -e HOME=/tmp -e USER=sam -itw /mnt --rm -v /home/sam/git/github/ray:/mnt python:3.9 /bin/bash

Run Ray linter.

python3 -m venv /tmp/venv
source /tmp/venv/bin/activate
./ci/lint/lint.sh code_format

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman committed:
Lint fix: Fix missing import and warnings

Run Python 3.9 docker container in a safe way.

docker run -name ray-lint -u 1000:1000 -e HOME=/tmp -e USER=sam -itw /mnt --rm -v /home/sam/git/github/ray:/mnt python:3.9 /bin/bash

Log into the container as root and install Java and Maven.

docker exec -itu root ray-lint /bin/bash

apt update
apt install -y openjdk-17-jdk maven

Run Ray linter from non-root container shell.

python3 -m venv /tmp/venv
source /tmp/venv/bin/activate
./ci/lint/lint.sh code_format

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

samrocketman commented:
Out of draft and the Python and Java parts are ready for review.  I'll push another commit for the C++ changes.

xieus commented:
@samrocketman great PR! I’m really glad to see this PR gaining traction and paving the way for native IPv6 support in Ray!

@anyscalesam Sam, is full IPv6 support included in Q1 2025 planning? :-)

samrocketman commented:
An update from me: this week I should be good to make the CPP/C changes after some other tasks.

Jan-M commented:
Great to see IPv6 getting attention and someone trying to solve it.

thundergolfer commented:
What's the intended `ray start` interface? I think by default `start` will look at IPv4 and pick it, so do you pass `--address`? In case it's interesting, this is the current behavior: 

```
root ~ → ray start --head --address ray://fdaa:ac1e:5066:5137:3ebf:f88:cc9e:1736:6379
Usage stats collection is enabled. To disable this, add `--disable-usage-stats` to the command that starts the cluster, or run the following command: `ray disable-usage-stats` before starting the cluster. See https://docs.ray.io/en/master/cluster/usage-stats.html for more details.

Specifying --address for external Redis address is deprecated. Please specify environment variable RAY_REDIS_ADDRESS=ray://fdaa:ac1e:5066:5137:3ebf:f88:cc9e:1736:6379 instead.
`--redis-username` should not be specified as empty string if external Redis server(s) `--address` points to requires username.
`--redis-password` should not be specified as empty string if external redis server(s) `--address` points to requires password.
Local node IP: 172.20.3.49
Traceback (most recent call last):
  File "/usr/local/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('ray', 'console_scripts', 'ray')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ray/python/ray/scripts/scripts.py", line 2708, in main
    return cli()
           ^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1082, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1697, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 788, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ray/python/ray/autoscaler/_private/cli_logger.py", line 823, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/root/ray/python/ray/scripts/scripts.py", line 890, in start
    node = ray._private.node.Node(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ray/python/ray/_private/node.py", line 167, in __init__
    maybe_key = self.check_persisted_session_name()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/ray/python/ray/_private/node.py", line 370, in check_persisted_session_name
    redis_ip_address, redis_port, enable_redis_ssl = get_address(
                                                     ^^^^^^^^^^^^
  File "/root/ray/python/ray/_private/services.py", line 1441, in get_address
    raise ValueError(
ValueError: Invalid redis address ray://fdaa:ac1e:5066:5137:3ebf:f88:cc9e:1736:6379.Expected format is ip:port or redis://ip:port, or rediss://ip:port for SSL.
```

samrocketman commented:
@thundergolfer your error in that area of code expects the URL to start with `redis://` or `rediss://`.

Also note, this PR is not quite ready for IPv6 testing.  It requires some C/C++ changes.  This PR is ready for testing IPv4 continuing to work as expected.

samrocketman commented:
I am also open to collaborating on writing automated tests for IPv6.  I have been monitoring PRs to this project and last time I resolved conflicts I needed to update code which was added using python sockets.

In order for future support of IPv6 to be stable, automated tests are necessary and people reviewing code would need to be cognizant of requesting changes in socket code not covered by IPv6 tests should they be written.

samrocketman committed:
Update address parsing for multiple ipv6 formats

IPv6 formats being `::` or `[::]`.

I used the following basic C++ to determine proper positioning of the index.

    #include <iostream>
    #include <string>
    using namespace std;
    int main() {
      const string gcs_address = "[2001:db8:3333:4444:5555:6666:7777:8888]:8443";
      size_t pos = gcs_address.rfind(':');
      int gcs_port_ = stoi(gcs_address.substr(pos + 1));
      cout << gcs_address << endl;
      cout << gcs_address[0] << endl;
      cout << gcs_address[pos - 1] << endl;
      cout << gcs_address.substr(1, pos - 2) << endl;
      cout << gcs_port_ << endl;
      return 0;
    }

Signed-off-by: Sam Gleske <sam.mxracer@gmail.com>
Signed-off-by: Sam Gleske <sgleske@integralads.com>
Co-authored-by: Sam Gleske <sgleske@integralads.com>

sgleske-ias commented:
C++ changes pulled from https://github.com/ray-project/ray/pull/40332, conflicts resolved, C++ code updated for my feature.

## This is ready for IPv6 testing

🎉 For those who operate IPv6 clusters please test this change.  I will push an update with master branch resolving conflicts later so that this PR is mergeable.

sgleske-ias commented:
Preserved original C++ authorship following this comment https://github.com/ray-project/ray/pull/44252#issuecomment-2089124798

sgleske-ias commented:
Here's an excerpt from one of the failures.

<details><summary>Failure stack trace (Click to expand)</summary>

```log

[2025-02-13T18:14:28Z] python/ray/tests/test_ray_init.py::test_ray_init_using_hostname [2025-02-13 18:14:27,789 E 75666 75666] logging.cc:111: Unhandled excepti
--
  | on: N5boost10wrapexceptINS_6system12system_errorEEE. what(): Invalid argument [system:22]
  | [2025-02-13T18:14:28Z] [2025-02-13 18:14:27,819 E 75666 75666] logging.cc:118: Stack trace:
  | [2025-02-13T18:14:28Z]  /rayci/python/ray/_raylet.so(+0x12c4c68) [0x7f1688219c68] ray::operator<<()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(+0x12c4e31) [0x7f1688219e31] ray::TerminateHandler()
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/../lib/libstdc++.so.6(+0xb135a) [0x7f16892d635a] __cxxabiv1::__terminate()
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/../lib/libstdc++.so.6(+0xb13c5) [0x7f16892d63c5]
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/../lib/libstdc++.so.6(+0xb1658) [0x7f16892d6658]
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(+0x1332e11) [0x7f1688287e11] boost::throw_exception<>()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(+0x1332d9a) [0x7f1688287d9a] boost::asio::detail::do_throw_error()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(+0x1333e86) [0x7f1688288e86]
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(_ZN3ray3rpc12BuildChannelERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiSt8optionalIN4grpc16Channel
  | ArgumentsEE+0x200) [0x7f168789a690] ray::rpc::BuildChannel()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(_ZN3ray3rpc12GcsRpcClient16CreateGcsChannelERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEi+0x1b7) [
  | 0x7f1687b05f37] ray::rpc::GcsRpcClient::CreateGcsChannel()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(_ZN3ray3rpc12GcsRpcClientC2ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiRNS0_17ClientCallManagerE
  | +0x10d) [0x7f1687b0476d] ray::rpc::GcsRpcClient::GcsRpcClient()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(_ZN3ray3gcs9GcsClient7ConnectER23instrumented_io_contextl+0xf8) [0x7f1687afe308] ray::gcs::GcsClient::Connec
  | t()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(_ZN3ray3gcs27ConnectOnSingletonIoContextERNS0_9GcsClientEl+0x45) [0x7f1687aff9c5] ray::gcs::ConnectOnSinglet
  | onIoContext()
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(+0x7f427b) [0x7f168774927b] __pyx_pw_3ray_7_raylet_14InnerGcsClient_1standalone()
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x507767] cfunction_call
  | [2025-02-13T18:14:28Z] /rayci/python/ray/_raylet.so(+0x831570) [0x7f1687786570] __pyx_tp_new_3ray_7_raylet_GcsClient()
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_MakeTpCall+0x1fb) [0x4f068b] _PyObject_MakeTpCall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x542c) [0x4ece2c] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4f81d3] function_code_fastcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x68f) [0x4e808f] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4f81d3] function_code_fastcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x68f) [0x4e808f] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4f81d3] function_code_fastcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x68f) [0x4e808f] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyFunction_Vectorcall+0xd5) [0x4f7ef5] _PyFunction_Vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_FastCallDictTstate+0x1a5) [0x4f0055] _PyObject_FastCallDictTstate
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x50281f] slot_tp_init
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_MakeTpCall+0x303) [0x4f0793] _PyObject_MakeTpCall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x542c) [0x4ece2c] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x50547d] method_vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(PyObject_Call+0xb4) [0x505bb4] PyObject_Call
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x44a1) [0x4ebea1] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyFunction_Vectorcall+0xd5) [0x4f7ef5] _PyFunction_Vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_FastCallDictTstate+0x1a5) [0x4f0055] _PyObject_FastCallDictTstate
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x50281f] slot_tp_init
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_MakeTpCall+0x303) [0x4f0793] _PyObject_MakeTpCall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x542c) [0x4ece2c] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyFunction_Vectorcall+0xd5) [0x4f7ef5] _PyFunction_Vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(PyObject_Call+0xb4) [0x505bb4] PyObject_Call
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x44a1) [0x4ebea1] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyFunction_Vectorcall+0xd5) [0x4f7ef5] _PyFunction_Vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x44a1) [0x4ebea1] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyFunction_Vectorcall+0xd5) [0x4f7ef5] _PyFunction_Vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x4d0f) [0x4ec70f] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4f81d3] function_code_fastcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x505371] method_vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x4d0f) [0x4ec70f] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4e6afa] _PyEval_EvalCode
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyFunction_Vectorcall+0xd5) [0x4f7ef5] _PyFunction_Vectorcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_FastCallDictTstate+0x1a5) [0x4f0055] _PyObject_FastCallDictTstate
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_Call_Prepend+0x66) [0x502de6] _PyObject_Call_Prepend
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x5cb733] slot_tp_call
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyObject_MakeTpCall+0x2ec) [0x4f077c] _PyObject_MakeTpCall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x542c) [0x4ece2c] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4f81d3] function_code_fastcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x68f) [0x4e808f] _PyEval_EvalFrameDefault
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3() [0x4f81d3] function_code_fastcall
  | [2025-02-13T18:14:28Z] /opt/miniconda/bin/python3(_PyEval_EvalFrameDefault+0x44a1) [0x4ebea1] _PyEval_EvalFrameDefault


```

</details>

Issues with `ray::gcs::GcsClient::Connect()`; I'll need to investigate.  Pull requests to my branch welcome if the fix is obvious to someone else.

thundergolfer commented:
@sgleske-ias what's your cluster testing infra like where you're able to run `python/ray/tests/test_ray_init.py::test_ray_init_using_hostname` and record a failure?

samrocketman commented:
@thundergolfer the excerpt I shared is from the failed GitHub check in this pull request.

laban-eilers-simplisafe commented:
@samrocketman We're very interested in this PR- we run IPv6-only EKS (kubernetes) clusters, and want to be able to run KubeRay on them. If there's anything we can do to help move this along, please let me know!

laban-eilers-simplisafe commented:
> @laban-eilers-simplisafe if you're familiar with c++ that's where I could use help at the moment;

Alas, I am not a C++ slinger. I will ask some other folks in my work who are to take a look, though.

juliencamille-hartmann-simplisafe commented:
Hey @sgleske-ias, hey samrocketman!
Thanks for the effort you have put in this. I have been reviewing it and I'd like to understand how we can help move things forward on the C++ side.

To help us scope our support, could you clarify the specific issues in the C++ changes where you need help?
And, to set expectations on our end:
- we would expect that C++ contributions are validated with minimal testing and aligned with Ray project contribution standard.

Happy to take a look once we have more clarity. Thanks again!

benjiebob commented:
+1 on this, it's blocking for us on our EKS cluster which only supports ipv6.

samrocketman commented:
@juliencamille-hartmann-simplisafe said:
> Hey @sgleske-ias, hey samrocketman! Thanks for the effort you have put in this. I have been reviewing it and I'd like to understand how we can help move things forward on the C++ side.
> 
> To help us scope our support, could you clarify the specific issues in the C++ changes where you need help? And, to set expectations on our end:
> 
>     * we would expect that C++ contributions are validated with minimal testing and aligned with Ray project contribution standard.
> 
> 
> Happy to take a look once we have more clarity. Thanks again!

Thanks for looking.  To put simply I wanted the tests to pass.  After I pushed C++ changes tests no longer pass.  So help figuring that out would be good.  I don't mind updating this PR to resolve conflicts again.

This area needs to be "all green" for it to be mergeable.  Beyond that, once all of the tests pass I defer to the project for what they consider mergeable.

![image](https://github.com/user-attachments/assets/f628c1b1-57f1-4ddf-83e6-1e8e9efe234f)

juliencamille-hartmann-simplisafe commented:
Hey again,

We took a closer look. Unfortunately, it seems to go much deeper than a localized fix.
While your change correctly parses the address as IPv6, it gets rejected shortly after. It turns out the broader Ray codebase assumes IPv4 in many places, and parsing is only the beginning. Changes are probably needed in most services, as well as supporting code (SSL...)

So, from what I've seen, supporting IPv6 across Ray would require a more comprehensive initiative, which is quite a bit more effort than we can realistically take on at this point.

It might make sense to raise this with the Ray maintainers, see if IPv6 support could be prioritized or scoped out officially.

Thanks again for kicking this off and digging in!

samrocketman commented:
> We took a closer look. Unfortunately, it seems to go much deeper than a localized fix. While your change correctly parses the address as IPv6, it gets rejected shortly after. It turns out the broader Ray codebase assumes IPv4 in many places, and parsing is only the beginning. Changes are probably needed in most services, as well as supporting code (SSL...)

Thanks for taking a look. I will likely continue chasing all of the broader changes in this pull request and continue to iterate until I succeed.

> So, from what I've seen, supporting IPv6 across Ray would require a more comprehensive initiative, which is quite a bit more effort than we can realistically take on at this point.

That's okay I appreciate your review.

> It might make sense to raise this with the Ray maintainers, see if IPv6 support could be prioritized or scoped out officially.

I think, in general, this is a hard requirement.  Since I initially opened this pull request there have been a few merged PRs which added socket code which would have broken IPv6 if it was hypothetically available.

Because of this, there needs to be supporting tests and an knowlege-sharing initiative for maintainers to actively preserve IPv6 support with new contributions.  I am happy to produce some sort of written guide or whatever written form makes sense for maintainers. I mentioned this in an earlier comment but I am reiterating it here.

At some point, I'll come back to this and dig in again (I'm away on vacation at the moment and this change start to finish has been spare time with exception for updating the PR once).

samrocketman commented:
@jjyao can you clarify what adding that label means?

samrocketman commented:
No stale; maintainers please label so stale bot does not mark as stale.

See also my previous questions

samrocketman commented:
To be clear I do not plan to resolve conflicts or work on this further until some maintainers come out of the wood work to talk with me about these changes.  I've requested multiple times in this pull request for a maintainer to chime in and there appears to be silence.

Because of this, I do not find it is worth my time to keep resolving conflicts with no maintainer conversation my time is not free.

You're welcome to reach out to me in this pull request or you may advise another method that's preferred.

robertnishihara commented:
@samrocketman thanks a ton for starting this work, and sorry about neglecting this!

I think we should do a quick call (or email / Slack) discussion to figure out the path forward. Some folks at ByteDance recently mentioned they have an implementation of this feature as well and are interested in contributing it.

Could you email me at robertnishihara@gmail.com or message me on the Ray Slack to have a higher bandwidth conversation? FYI @richardliaw @jjyao

samrocketman commented:
@robertnishihara emailed

samrocketman commented:
Update: today at 2pm ET there was an in-person meeting between maintainers and I where we went over the current pull request and general information on how to support IPv6.

Much of what I've already shared in this pull request was shared within the meeting.

psaab commented:
You need this diff as well or something similar

``diff --git a/ray/python/ray/dashboard/modules/dashboard_sdk.py b/ray/python/ray/dashboard/modules/dashboard_sdk.py
--- a/ray/python/ray/dashboard/modules/dashboard_sdk.py
+++ b/ray/python/ray/dashboard/modules/dashboard_sdk.py
@@ -298,7 +298,13 @@
         Keyword arguments other than "cookies", "headers" are forwarded to the
         `requests.request()`.
         """
-        url = self._address + endpoint
+        url = f"{self._address}{endpoint}"
+        try:
+            if ipaddress.ip_address(self._address).version == 6:
+                url = f"[{self._address}]{endpoint}"
+        except ValueError:
+            pass
+``

samrocketman commented:
@psaab mind opening a PR to my fork so that you get attributed for that diff?
